rcl/src/rcl/wait.c

```diff
@@ -614,6 +610,157 @@ rcl_wait(rcl_wait_set_t * wait_set, int64_t timeout)
   return RCL_RET_OK;
 }
 
+rcl_ret_t
+rcl_wait_multiple(rcl_wait_set_t wait_sets[], const size_t num_wait_sets, int64_t timeout)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(wait_sets, RCL_RET_INVALID_ARGUMENT);
+  if (0u == num_wait_sets) {
+    RCL_SET_ERROR_MSG("number of wait sets must be greater than zero");
+    return RCL_RET_INVALID_ARGUMENT;
+  }
+
+  // Count number of entities in wait sets while checking for invalid wait sets
+  size_t number_of_subscriptions = 0u;
+  size_t number_of_guard_conditions = 0u;
+  size_t number_of_timers = 0u;
+  size_t number_of_clients = 0u;
+  size_t number_of_services = 0u;
+  for (size_t i = 0u; i < num_wait_sets; ++i) {
+    rcl_wait_set_t * wait_set = &wait_sets[i];
+    if (!__wait_set_is_valid(wait_set)) {
+      RCL_SET_ERROR_MSG("wait set is invalid");
+      return RCL_RET_WAIT_SET_INVALID;
+    }
+    number_of_subscriptions += wait_set->size_of_subscriptions;
+    number_of_guard_conditions += wait_set->size_of_guard_conditions;
+    number_of_timers += wait_set->size_of_timers;
+    number_of_clients += wait_set->size_of_clients;
+    number_of_services += wait_set->size_of_services;
+  }
+  if (
+    number_of_subscriptions == 0u &&
+    number_of_guard_conditions == 0u &&
+    number_of_timers == 0u &&
+    number_of_clients == 0u &&
+    number_of_services == 0u)
+  {
+    RCL_SET_ERROR_MSG("wait sets are empty");
+    return RCL_RET_WAIT_SET_EMPTY;
+  }
+
+  // Create a single wait set where all wait set entities can be collated
+  rcl_wait_set_t collated_wait_set = rcl_get_zero_initialized_wait_set();
+  rcl_ret_t ret = rcl_wait_set_init(
+    &collated_wait_set,
+    number_of_subscriptions,
+    number_of_guard_conditions,
+    number_of_timers,
+    number_of_clients,
+    number_of_services,
+    wait_sets[0].impl->allocator);
+  if (RCL_RET_OK != ret) {
+    if (RCL_RET_BAD_ALLOC == ret) {
+      return RCL_RET_BAD_ALLOC;  // error already set
+    }
+    return RCL_RET_ERROR;  // error already set
+  }
+
+  // Collate all wait sets into a single wait set
+  for (size_t i = 0u; i < num_wait_sets; ++i) {
+    rcl_wait_set_t * wait_set = &wait_sets[i];
+    for (size_t j = 0u; j < wait_set->size_of_timers; ++j) {
+      const size_t current_index = collated_wait_set.impl->timer_index++;
+      collated_wait_set.timers[current_index] = wait_set->timers[j];
+      rcl_guard_condition_t * guard_condition = rcl_timer_get_guard_condition(wait_set->timers[j]);
+      if (NULL != guard_condition) {
+        // rcl_wait() will take care of moving these backwards and setting guard_condition_count.
+        const size_t offset_rmw_index = wait_set->size_of_guard_conditions + current_index;
```

Comment:
`offset_rmw_index` is in the combined wait set, so I think this should be `collated_wait_set->size_of_guard_conditions ...`

Reply:
Yeah, good catch!

Reply:
Fixed in b58dfa9